### ISSUE TITLE
std.json failing test cases

### ISSUE THREAD

**Issue by andrewrk** [ziglang/zig#2379]
Reported by @shawnl in #2239

```zig
// https://tools.ietf.org/html/rfc8259#section-8.1
test "json.invalidutf8" {
    err("[\"\xff\x80\x80\x80invalid\"]");
    err("[\"\xc0\x80invalid\"]");
    err("[\"\xed\xa0short\"]");
    err("[\"\xf0\x80\x80\x80short\"]");
    err("[\"\xf4\x90\x80\x80long\"]");
    err("[\"\xc2trunc\"]");
    err("[\"\xee\x80trunc\"]");
    err("[\"\xf4\x80\x80trunc\"]");
}

// https://tools.ietf.org/html/rfc8259#section-8.2
test "json.invalidunicode" {
    err(
        \\["\uDEAD"]
    );
} 
```

**Comment by andrewrk** [comment]
Confirmed still an issue.

**Comment by hryx** [comment]
A clarification about that second test example (`\uDEAD`). It's a little nasty.

TLDR; I recommend we continue to accept `\uDEAD` until we have decode/encode options.

[Section 8.2](https://tools.ietf.org/html/rfc8259#section-8.2), mentioned in the above test, explicitly admits that the included grammar allows codepoints D800-DFFF, even though they are illegal UTF-8. It basically follows with "implementations may vary or crash, so use those characters at your own risk". Windows file paths, WTF-8, blah blah blah.

But [section 8.1](https://tools.ietf.org/html/rfc8259#section-8.1) says that if you exchange JSON "between systems that are not part of a closed ecosystem", then you have to use real UTF-8 (i.e. disallow `\uDEAD` etc.). I would assume this is the safest approach when dealing with third-party APIs and so on.

Anyway, thanks @daurnimator for helping me sort out those nuances on IRC. Daurnimator suggests encode/decode options, which sounds appropriate to me. If you're sending JSON to any arbitrary HTTP API, you likely want to encode (or accept) according to the stricter spec.

That first test case contains definitively invalid sequences. I have a fix which I'll try to push tonight.
